File: src/eval/scala/org/dbpedia/spotlight/evaluation/DBpediaSpotlightClient.scala
/**
 * Copyright 2011 Pablo Mendes, Max Jakob
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.dbpedia.spotlight.evaluation

import org.apache.commons.logging.LogFactory
import java.io.{PrintStream, File}
import io.Source
import org.dbpedia.spotlight.annotate.DefaultAnnotator
import org.dbpedia.spotlight.util.AnnotationFilter
import org.dbpedia.spotlight.string.WikiLinkParser
import scala.collection.JavaConversions._
import org.dbpedia.spotlight.model.{DBpediaResource, DBpediaResourceOccurrence}
/**
 * Reads in manually annotated paragraphs, computes the inter-annotator agreement, then compares
 * our system against the union or intersection of the manual annotators.
 *
 * TODO Create client for Spotlight in the same style created for Alchemy, Ontos, OpenCalais, WikiMachine and Zemanta. i.e. using the Web Service.
 *
 * @author pablomendes
 */
object DBpediaSpotlightClient
{
  private val LOG = LogFactory.getLog(this.getClass)

  val spotterFile    = "/home/pablo/web/TitRedDis.spotterDictionary";
  //val spotterFile    = "/home/pablo/eval/manual/Eval.spotterDictionary";
  val indexDirectory = "/home/pablo/web/DisambigIndex.singleSFs-plusTypes.SnowballAnalyzer.DefaultSimilarity";
  //val indexDirectory = "/home/pablo/web/DisambigIndex.restrictedSFs.plusTypes-plusSFs.SnowballAnalyzer.DefaultSimilarity"
  val confidence = 0.0;
  val support = 0;

  val targetTypesList = null;
  val coreferenceResolution = true;
  val annotator = new DefaultAnnotator(new File(spotterFile), new File(indexDirectory));

  def parseToMatrix(occList : List[DBpediaResourceOccurrence]) : String = {
    val buffer = new StringBuffer();
    var pos = 0;
    var text = "";
    for (occ <- occList) {
      text = occ.context.text;
      val chunk = occ.context.text.substring(pos, occ.textOffset);
      WikiLinkParser.appendToMatrix(chunk, new DBpediaResource(WikiLinkParser.NoTag), buffer);
      WikiLinkParser.appendToMatrix(occ.surfaceForm.name, occ.resource, buffer);
      pos = occ.textOffset + occ.surfaceForm.name.length();
    }
    WikiLinkParser.appendToMatrix(text.substring(pos),new DBpediaResource(WikiLinkParser.NoTag),buffer);
    return buffer.toString();
  }

  def main(args : Array[String])
    {
      
//      val baseDir: String = "/home/pablo/eval/manual"
//      val inputFile: File = new File(baseDir+"AnnotationText.txt");
//      val prefix = "spotlight/AnnotationText-Spotlight";
//      val outputFile: File = new File(baseDir+prefix+".txt.matrix");

//      val baseDir: String = "/home/pablo/eval/cucerzan/"
//      val inputFile: File = new File(baseDir+"cucerzan.txt");

      val baseDir: String = "/home/pablo/eval/manual/"
      val inputFile: File = new File(baseDir+"AnnotationText.txt");

      val prefix = "spotlight/Spotlight";
      val setOutputFile: File = new File(baseDir+prefix+"NoFilter.set");
      val matrixOutputFile = new File(baseDir+prefix+".matrix");
      if (!new File(baseDir).exists) {
        System.err.println("Base directory does not exist. "+baseDir);
        exit();
      }
      
      //ANNOTATE AND TRANSFORM TO MATRIX

      val plainText = Source.fromFile(inputFile).mkString // Either get from file
      //val plainText = AnnotatedTextParser.eraseMarkup(text)     //   or erase markup

      val top10Score = new PrintStream(baseDir+prefix+"Top10Score.set");
      val top10Confusion = new PrintStream(baseDir+prefix+"Top10Confusion.set");
      val top10Prior = new PrintStream(baseDir+prefix+"Top10Prior.set");
      val top10Confidence = new PrintStream(baseDir+prefix+"Top10Confidence.set");
      val top10TrialAndError = new PrintStream(baseDir+prefix+"Top10TrialAndError.set");
      //val out = new PrintStream(matrixOutputFile);
      var occurrences = List[DBpediaResourceOccurrence]();
        for (text <- plainText.split("\n")) {
          val cleanText = WikiLinkParser.eraseMarkup(text);          

          val occList = annotator.annotate(cleanText);
          val occ = occList.toList
          //out.append(parseToMatrix(occ));
          occurrences = occurrences ::: occ;

          //val sixPercent = (cleanText.split("\\s+").size * 0.06).round.toInt;
          val sixPercent = 20;
          val k = Math.min(occ.size, sixPercent)
          
          val entitiesByScore = occ.sortBy(o=>o.similarityScore).reverse.map(o=>o.resource.uri);
          top10Score.append(entitiesByScore.slice(0,k).toSet.mkString("\n"))
          top10Score.append("\n");

          val entitiesByConfusion = occ.sortBy(o=>o.percentageOfSecondRank).map(o=>o.resource.uri); //should not be reversed. small percentage is large gap.
          top10Confusion.append(entitiesByConfusion.slice(0,k).toSet.mkString("\n"))
          top10Confusion.append("\n");

          val entitiesByPrior = occ.sortBy(o=>o.spotProb).reverse.map(o=>o.resource.uri);
          top10Prior.append(entitiesByPrior.slice(0,k).toSet.mkString("\n"))
          top10Prior.append("\n");

          val entitiesByConfidence = occ.sortBy(o=> (o.similarityScore * (1-o.percentageOfSecondRank)) ).reverse.map(o=>o.resource.uri);
          top10Confidence.append(entitiesByConfidence.slice(0,k).toSet.mkString("\n"))
          top10Confidence.append("\n");

          val entitiesByTrialAndError = occ.sortBy(o=> (o.similarityScore * (1-o.percentageOfSecondRank) * o.spotProb) ).reverse.map(o=>o.resource.uri);
          top10TrialAndError.append(entitiesByTrialAndError.slice(0,k).toSet.mkString("\n"))
          top10TrialAndError.append("\n");

        }
      //out.close();
      top10Score.close()
      top10Prior.close()
      top10Confusion.close()
      top10Confidence.close()
      top10TrialAndError.close()


      //val occList : java.util.List[DBpediaResourceOccurrence] = occurrences
      //writeAsEntitySetToFile(occurrences, setOutputFile)
      val allOut = new PrintStream(setOutputFile)
      allOut.append(occurrences.map(occ => occ.resource.uri).toSet.mkString("\n"))
      allOut.close();

//      val top6pc = (0.6 * occurrences.size).toInt
//      val entitiesByPercentageRank = occurrences.map(o => (o.resource,o.percentageOfSecondRank)).distinct.sortBy(t=>t._2);
//      writeAsEntitySetToFile(entitiesByScore.slice(0, top6pc), new File(baseDir+prefix+".sim.t6pc.set"))
//      writeAsEntitySetToFile(entitiesByPercentageRank.slice(0, top6pc), new File(baseDir+prefix+".rank.t6pc.set"))
//      writeAsEntitySetToFile(entitiesByPercentageRank.slice(0,10), new File(baseDir+prefix+".rank.t10.set"))


      val filteredOccList : List[DBpediaResourceOccurrence] = AnnotationFilter.filter(occurrences, 0, 0, AnnotationFilter.DEFAULT_TYPES, "", false, AnnotationFilter.DEFAULT_COREFERENCE_RESOLUTION);
      //val filteredOccList : List[DBpediaResourceOccurrence] = AnnotationFilter.filter(occurrences, 0, 0, AnnotationFilter.PERLOCPLA, AnnotationFilter.DEFAULT_COREFERENCE_RESOLUTION);
      for (confidence <- EvalParams.confidenceInterval) {
          for(support <- EvalParams.supportInterval) {
            var localFiltered = AnnotationFilter.filterBySupport(filteredOccList, support)
            localFiltered = AnnotationFilter.filterByConfidence(localFiltered, confidence)
            val out = new PrintStream(new File(baseDir+prefix+".c"+confidence+"s"+support+".set"))
            out.append(localFiltered.map(occ => occ.resource.uri).toSet.mkString("\n"))
            out.close();
          }
      }

    }

  def writeAsEntitySetToFile(filteredOccList: java.util.List[DBpediaResourceOccurrence], file: File) {
    val out = new PrintStream(file)
    val entitySet = filteredOccList
    out.append(entitySet.map(occ => occ.resource.uri).toSet.mkString("\n"))
    out.close();
  }

    def writeAsEntitySetToFile(occList: List[Tuple2[DBpediaResource,Double]], file: File) {
    val out = new PrintStream(file)
    out.append(occList.map( t => t._1.uri).toSet.mkString("\n"))
    out.close();
  }

}

File: eval/src/main/java/org/dbpedia/spotlight/evaluation/external/DBpediaSpotlightClient.java
/**
 * Copyright 2011 Pablo Mendes, Max Jakob
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.dbpedia.spotlight.evaluation.external;

import org.apache.commons.httpclient.Header;
import org.apache.commons.httpclient.methods.GetMethod;
import org.dbpedia.spotlight.exceptions.AnnotationException;
import org.dbpedia.spotlight.model.DBpediaResource;
import org.dbpedia.spotlight.model.Text;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.LinkedList;
import java.util.List;

/**
 * Simple web service-based annotation client for DBpedia Spotlight.
 *
 * @author pablomendes, Joachim Daiber
 */

public class DBpediaSpotlightClient extends AnnotationClient {

	//private final static String API_URL = "http://jodaiber.dyndns.org:2222/";
    private final static String API_URL = "http://spotlight.dbpedia.org/";
	private static final double CONFIDENCE = 0.0;
	private static final int SUPPORT = 0;

	@Override
	public List<DBpediaResource> extract(Text text) throws AnnotationException {

        LOG.info("Querying API.");
		String spotlightResponse;
		try {
			GetMethod getMethod = new GetMethod(API_URL + "rest/annotate/?" +
					"confidence=" + CONFIDENCE
					+ "&support=" + SUPPORT
					+ "&text=" + URLEncoder.encode(text.text(), "utf-8"));
			getMethod.addRequestHeader(new Header("Accept", "application/json"));

			spotlightResponse = request(getMethod);
		} catch (UnsupportedEncodingException e) {
			throw new AnnotationException("Could not encode text.", e);
		}

		assert spotlightResponse != null;

		JSONObject resultJSON = null;
		JSONArray entities = null;

		try {
			resultJSON = new JSONObject(spotlightResponse);
			entities = resultJSON.getJSONArray("Resources");
		} catch (JSONException e) {
			throw new AnnotationException("Received invalid response from DBpedia Spotlight API.");
		}

		LinkedList<DBpediaResource> resources = new LinkedList<DBpediaResource>();
		for(int i = 0; i < entities.length(); i++) {
			try {
				JSONObject entity = entities.getJSONObject(i);
				resources.add(
						new DBpediaResource(entity.getString("@URI"),
								Integer.parseInt(entity.getString("@support"))));

			} catch (JSONException e) {
                LOG.error("JSON exception "+e);
            }

		}


		return resources;
	}
        public static void main(String[] args) throws Exception {


        DBpediaSpotlightClient c = new DBpediaSpotlightClient ();

//        File input = new File("/home/pablo/eval/manual/AnnotationText.txt");
//        File output = new File("/home/pablo/eval/manual/systems/Spotlight.list");

        //File input = new File("/home/pablo/eval/cucerzan/cucerzan.txt");
        //File output = new File("/home/pablo/eval/cucerzan/systems/cucerzan-Spotlight.set");

//        File input = new File("/home/pablo/eval/wikify/gold/WikifyAllInOne.txt");
//        File output = new File("/home/pablo/eval/wikify/systems/Spotlight.list");

        File input = new File("/home/alexandre/Projects/test-files-spotlight/ExternalClients_TestFiles/Berlin.txt");
        File output = new File("/home/alexandre/Projects/test-files-spotlight/ExternalClients_TestFiles/Spotlight.list");


        c.evaluate(input, output);


//        SpotlightClient c = new SpotlightClient(api_key);
//        List<DBpediaResource> response = c.extract(new Text(text));
//        PrintWriter out = new PrintWriter(manualEvalDir+"AnnotationText-Spotlight.txt.set");
//        System.out.println(response);

    }



}
